README.md

[![Build status](https://ci.appveyor.com/api/projects/status/4uaa2irr26deoffv?svg=true)](https://ci.appveyor.com/project/chrismetcalf/soda-ruby)
[![Build Status](https://travis-ci.org/socrata/soda-ruby.svg?branch=master)](https://travis-ci.org/socrata/soda-ruby)

For more details and for documentation, check out <http://socrata.github.io/soda-ruby> or our [developer portal](http://dev.socrata.com).

## Installation

SODA is distributed as a gem, which is how it should be used in your app.

Include the gem and hashie in your Gemfile:

```ruby
gem 'soda-ruby', :require => 'soda'
```

### Important Note!

In order to access the SODA API via HTTPS, clients must now [support the Server Name Indication (SNI)](https://dev.socrata.com/changelog/2016/08/24/sni-now-required-for-https-connections.html) extension to the TLS protocol. What does this mean? It means that if you're using `soda-ruby`, you must [use Ruby 2.0.0 or above](https://en.wikipedia.org/wiki/Server_Name_Indication), as that is when `net/http` introduced support for SNI. 2.0.0 was released in 2011, so most up-to-date platforms will be on version 2.0 or greater.

## Quick Start

Create a new client. Register for an application token at <http://dev.socrata.com/register>.

```ruby
client = SODA::Client.new({:domain => "explore.data.gov", :app_token => "CGxadgoQlgQSev4zyUh5aR5J3"})
```

Issue a filter query. `644b-gaut` is the identifier for the dataset we want to access.

### As of version 1.0.0+

The return object is the complete response object with a pre parsed body. The response.body object is an array of [Hashie::Mash].

If you are upgrading from a version < 1.0.0 The previous object returned is now the response.body object.

### Prior to version 1 (<1.0.0)

The return object is an array of [Hashie::Mash] that represents the body of the response.

(https://github.com/intridea/hashie) objects:

```ruby
response = client.get("644b-gaut", {"$limit" => 1, :namelast => "WINFREY", :namefirst => "OPRAH"})

 #=> [#<Hashie::Mash appt_end_date="12/3/09 23:59" appt_made_date="12/2/09 18:05" appt_start_date="12/3/09 9:30" caller_name_first="SALLY" caller_name_last="ARMBRUSTER" last_updatedby="J7" lastentrydate="12/2/09 18:05" meeting_loc="WH" meeting_room="DIP ROOM" namefirst="OPRAH" namelast="WINFREY" post="WIN" release_date=1269586800 terminal_suffix="J7" total_people="10" type_of_access="VA" uin="U60816" visitee_namefirst="SEMONTI" visitee_namelast="STEPHENS">]
```

You can use other simple query SoQL methods found here: <http://dev.socrata.com/docs/queries.html>

```ruby
client = SODA::Client.new({:domain => "soda.demo.socrata.com"})
magnitude_response = client.get("4tka-6guv", {"$where" => "magnitude > '3.0'"})

#=> [#<Hashie::Mash datetime="2012-09-14T09:28:55" depth="20" earthquake_id="12258012" location=#<Hashie::Mash latitude="19.7859" longitude="-64.0849" needs_recoding=false> magnitude="3.1" number_of_stations="6" region="north of the Virgin Islands" source="pr" version="0">, #<Hashie::Mash datetime="2012-09-14T07:58:39" depth="74" earthquake_id="12258011" location=#<Hashie::Mash latitude="19.5907" longitude="-64.1723" needs_recoding=false> magnitude="3.3" number_of_stations="4" region="Virgin Islands region" source="pr" version="0">, ... ]

datetime_response = client.get("4tka-6guv", {"$where" => "datetime > '2012-09-14T09:28:55' AND datetime < '2012-12-25T09:00:00'"})

#=> [#<Hashie::Mash datetime="2012-09-14T10:10:19" depth="8.2" earthquake_id="00388609" location=#<Hashie::Mash latitude="36.9447" longitude="-117.6778" needs_recoding=false> magnitude="1.7" number_of_stations="29" region="Northern California" source="nn" version="9">, #<Hashie::Mash datetime="2012-09-14T10:06:11" depth="6.4" earthquake_id="00388607" location=#<Hashie::Mash latitude="36.9417" longitude="-117.6903" needs_recoding=false> magnitude="1.7" number_of_stations="29" region="Central California" source="nn" version="9">, ... ]
```

You can also provide a full URI to an API endpoint instead of specifying the ID. Just copy and paste the dataset URI from the API documentation!

```ruby
client = SODA::Client.new({:domain => "soda.demo.socrata.com"})
magnitude_response = client.get("https://soda.demo.socrata.com/resource/4tka-6guv.json", {"$where" => "magnitude > '3.0'"})
```

All the field names have built in getter methods since the objects are Hashie::Mashes.

```ruby
magnitude_response.first.number_of_stations   #=> "6"
```
*Note that the return value is a string object.*

## Contribution

### How to Publish
Raw ruby-gems publishing information can be found at [https://guides.rubygems.org/publishing/](https://guides.rubygems.org/publishing/).

1) First you must publish with a `soda-ruby` maintainer account. Please use the `tylsvc-rubygems` rubygems.org account located in lasspass. If you do not see it in lasspass, please reach out to a lasspass admin in order to get access.


1) After finding the account, proceed to your local `soda-ruby` project directory. Update the **version** of the file project inside the `soda.gemspec` file, under `s.version = '1.0.1'`. This project uses semantic versioning, learn more here [https://semver.org/](https://semver.org/).


1) Now build the gem by running `gem build soda.gemspec`. It should output a file with the name `soda-ruby-VERSION_NUMBER.gem`.

2) Now to publish the gem run `gem publish soda-ruby-VERSION_NUMBER.gem`. Please insert the correct VERSION_NUMBER. This is where gems prompt for the `tylsvc-rubygems` user name and password.

Congrats! The gem should now be published.



CHANGELOG.md
0.2.25
------
- Bumping hashie to 3.5

0.2.24
------
- [#44](https://github.com/socrata/soda-ruby/pull/44) Relaxing strictness of hashie dependency

0.2.23
------
- [#42](https://github.com/socrata/soda-ruby/issues/42) Removing a spurious dependency on `activesupport`

0.2.22
------
- Fix `send_request` from sending wrong number of arguments to `delete_method_response` causing `client.delete` method to error out

0.2.21
------
Re-push of 0.2.20 because I got confused about how rubygems.org handled bad versions.

0.2.20
------
Internal cleanup and bug fixes

0.2.19
------
Added the ability to pass in a full URI instead of just a path or dataset identifier when making a request. You can also combine URIs with additional parameters.

Ex:

    require 'soda/client'
    client = SODA::Client.new
    client.get("https://fakehost.socrata.com/resource/644b-gaut.json?last_name=Willis", :first_name => "Bruce")

0.2.18
------
Added a `:debug_stream` option to the client to all you to get the full HTTP conversation for debugging. Should not be used in production.

    require 'soda/client'
    client = SODA::Client.new(:domain => "data.seattle.gov", :debug_stream => $stderr)

0.2.17
------
Modified `SODA::Client.generate_user_agent` to not use `Sys::Uname` to fetch system information.

0.2.16
------
- Added support for OAuth tokens for authentication
- Improved the detail of the exceptions we throw in the case of errors

0.2.15
------
(This version was never pushed)
- Added an OmniAuth provider

0.2.13
------
Added a global :timeout option to override the default Net:HTTP timeout, and accepting 202 as an "OK" response code, which should only occur during import.

0.2.12
------
Added a global option to disable SSL checking (#7)

0.2.11
------
We weren't properly handling an empty payload response

0.2.10
------
Removing a nesting limit that we missed

0.2.8
-----
Removing the nesting limit on the JSON parser

0.2.7
-----
Removing dead code and an un-needed dependency

0.2.6
-----
Code cleanup from @raykao

0.2.4
-----
Point release including tests that are now mocked properly

0.2.1
-----
Point release to update gemspec dependencies

0.2.0
-----
Initial Public Release



test/test_soda.rb
require 'test/unit'
require 'shoulda'
require 'soda/client'
require 'soda/exceptions'
require 'json'
require 'sys/uname'
require 'webmock/test_unit'
require 'mocha/test_unit'

# NOTE: These tests are by no means exhaustive, they're just a start
class SODATest < Test::Unit::TestCase
  DOMAIN = 'fakehost.socrata.com'
  APP_TOKEN = 'totallyfakenotrealapptoken'
  USER = 'fakeuser@socrata.com'
  PASSWORD = 'fakepassword'

  # Helpers
  def resource(name)
    File.new(File.dirname(__FILE__) + '/resources/' + name)
  end

  context 'user agents' do
    should 'return a proper user agent string for mac osx' do
    end
  end

  context 'query strings' do
    setup do
      @client = SODA::Client.new
    end

    should 'return a proper simple query string' do
      assert_equal 'foo=bar', @client.send(:query_string, :foo => 'bar')
    end

    should 'return a proper multi-param query string' do
      assert_equal 'foo=bar&quo=quux', @client.send(:query_string, :foo => 'bar',
                                                                   :quo => 'quux')
    end

    should 'escape spaces' do
      assert_equal 'foo=bar+has+spaces+in+it', @client.send(:query_string, :foo => 'bar has spaces in it')
    end

    should 'escape random symbols' do
      assert_equal 'foo=%21%40%23%24%25%5E%26%2A%28%29_%2B', @client.send(:query_string, :foo => '!@#$%^&*()_+')
    end
  end

  context 'resource paths with domain' do
    setup do
      @client = SODA::Client.new(:domain => 'fakehost.socrata.com')
    end

    should 'handle a simple resource' do
      assert_equal 'https://fakehost.socrata.com/resource/644b-gaut.json', @client.send(:parse_resource, '644b-gaut')
    end

    should 'handle a custom resource' do
      assert_equal 'https://fakehost.socrata.com/resource/visitor-records.json',
                   @client.send(:parse_resource, 'visitor-records')
    end

    should 'allow you to override content type' do
      assert_equal 'https://fakehost.socrata.com/resource/visitor-records.csv',
                   @client.send(:parse_resource, 'visitor-records.csv')
    end

    should 'allow you to pass in a full URL for another domain if that\'s your thing' do
      assert_equal 'https://anotherfakehost.socrata.com/resource/1234-abcd.csv',
                   @client.send(:parse_resource,
                                'https://anotherfakehost.socrata.com/resource/1234-abcd.csv')
    end

    # NOTE: Will be deprecated later
    should 'allow you access an old-style SODA1 path' do
      assert_equal 'https://fakehost.socrata.com/api/views/644b-gaut.json',
                   @client.send(:parse_resource, '/api/views/644b-gaut')
    end

    # NOTE: Will be deprecated later
    should 'allow you access an old-style SODA1 path with output type' do
      assert_equal 'https://fakehost.socrata.com/api/views/644b-gaut/rows.csv',
                   @client.send(:parse_resource, '/api/views/644b-gaut/rows.csv')
    end
  end

  context 'resource paths without domain' do
    setup do
      @client = SODA::Client.new
    end

    should 'fail if you don\'t provide a full URI' do
      assert_raise RuntimeError do
        @client.send(:parse_resource, '644b-gaut')
      end
    end

    should 'work with a full URI' do
      assert_equal 'https://anotherfakehost.socrata.com/resource/1234-abcd.csv',
                   @client.send(:parse_resource,
                                'https://anotherfakehost.socrata.com/resource/1234-abcd.csv')
    end
  end

  # Test our response handling directly
  context 'response objects' do
    setup do
      @client = SODA::Client.new(:domain => DOMAIN, :app_token => APP_TOKEN)
    end

    should 'handle a 200 with a real JSON payload' do
      stub_request(:get, 'https://fakehost.socrata.com/resource/earthquakes.json')
        .to_return(resource('earthquakes_50.response'))

      uri = URI.parse('https://fakehost.socrata.com/resource/earthquakes.json')
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true

      request = Net::HTTP::Get.new(uri.request_uri)
      request.add_field('X-App-Token', 'FAKEAPPTOKEN')

      body = @client.send(:handle_response, http.request(request))
      assert_equal 50, body.size
    end

    should 'handle a 200 with an empty payload' do
      stub_request(:get, 'https://fakehost.socrata.com/resource/empty.json')
        .to_return(resource('empty_body.response'))

      uri = URI.parse('https://fakehost.socrata.com/resource/empty.json')
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true

      request = Net::HTTP::Get.new(uri.request_uri)
      request.add_field('X-App-Token', 'FAKEAPPTOKEN')

      body = @client.send(:handle_response, http.request(request))
      assert_nil body
    end

    should 'raise on a 500 error' do
      stub_request(:get, 'https://fakehost.socrata.com/kaboom.json')
        .to_return(resource('500_error.response'))

      uri = URI.parse('https://fakehost.socrata.com/kaboom.json')
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true

      request = Net::HTTP::Get.new(uri.request_uri)
      request.add_field('X-App-Token', 'FAKEAPPTOKEN')

      assert_raise SODA::InternalServerError do
        @client.send(:handle_response, http.request(request))
      end

      # Confirm that our exceptions can still be caught by RuntimeError
      begin
        @client.send(:handle_response, http.request(request))
      rescue StandardError => e
        assert e.class <= Exception
      end
    end
  end

  # Test responses from a real fake dataset
  context 'earthquakes' do
    setup do
      SODA::Client.stubs(:generate_user_agent).returns('The Best User Agent of All Time')
      @client = SODA::Client.new(:domain => DOMAIN, :app_token => 'K6rLY8NBK0Hgm8QQybFmwIUQw')
    end

    should 'be able to access the earthquakes dataset' do
      stub_request(:get, 'https://fakehost.socrata.com/resource/earthquakes.json?')
        .with(:body => 'null',
              :headers => { 'Accept' => '*/*', 'Content-Type' => 'application/json',
                            'User-Agent' => 'The Best User Agent of All Time',
                            'X-App-Token' => 'K6rLY8NBK0Hgm8QQybFmwIUQw' })
        .to_return(resource('earthquakes_50.response'))

      response = @client.get('earthquakes')
      assert_equal 50, response.size
    end

    should 'be able to perform a simple equality query' do
      stub_request(:get, 'https://fakehost.socrata.com/resource/earthquakes.json?source=uw')
        .to_return(resource('earthquakes_uw.response'))

      response = @client.get('earthquakes', :source => 'uw')
      assert_equal 26, response.size
    end

    should 'be able to perform a simple $WHERE query' do
      stub_request(:get, 'https://fakehost.socrata.com/resource/earthquakes.json?$where=magnitude > 5')
        .to_return(resource('earthquakes_where_gt_5.response'))

      response = @client.get('earthquakes', '$where' => 'magnitude > 5')
      assert_equal 29, response.size
    end

    should 'be able to combine a $WHERE query with a simple equality' do
      stub_request(:get, 'https://fakehost.socrata.com/resource/earthquakes.json?$where=magnitude > 4&source=pr')
        .to_return(resource('earthquakes_source_pr_where_gt_4.response'))

      response = @client.get('earthquakes', '$where' => 'magnitude > 4', :source => 'pr')
      assert_equal 1, response.size
    end

    should 'get the results we expect' do
      stub_request(:get, 'https://fakehost.socrata.com/resource/earthquakes.json?$where=magnitude > 4&source=pr')
        .to_return(resource('earthquakes_source_pr_where_gt_4.response'))

      response = @client.get('earthquakes', '$where' => 'magnitude > 4', :source => 'pr')
      assert_equal 1, response.size

      quake = response.first
      assert_equal 'Puerto Rico region', quake.region
      assert_equal '4.2', quake.magnitude
      assert_equal '17.00', quake.depth

      assert quake.region?
    end

    should 'allow you to issue a query as a full URL string' do
      stub_request(:get, 'https://fakehost.socrata.com/resource/earthquakes.json?source=uw')
        .to_return(resource('earthquakes_uw.response'))

      response = @client.get('https://fakehost.socrata.com/resource/earthquakes.json?source=uw')
      assert_equal 26, response.size
    end

    should 'allow you to combine a full URL with an extra bit of query string' do
      stub_request(:get, 'https://fakehost.socrata.com/resource/earthquakes.json?$where=magnitude > 4&source=pr')
        .to_return(resource('earthquakes_source_pr_where_gt_4.response'))

      response = @client.get('https://fakehost.socrata.com/resource/earthquakes.json?source=pr',
                             '$where' => 'magnitude > 4')
      assert_equal 1, response.size
    end
  end

  context 'authenticated' do
    setup do
      SODA::Client.stubs(:generate_user_agent).returns('The Best User Agent of All Time')
      @client = SODA::Client.new(:domain => DOMAIN, :app_token => APP_TOKEN,
                                 :username => USER, :password => PASSWORD)
    end

    context 'with user accounts' do
      should 'be able to read his own email address' do
        stub_request(:get, 'https://fakehost.socrata.com/api/users/current.json')
          .with(:body => "null",
                :headers => {
                  'Accept' => '*/*',
                  'Accept-Encoding' => 'gzip;q=1.0,deflate;q=0.6,identity;q=0.3',
                  'Authorization' => 'Basic ZmFrZXVzZXJAc29jcmF0YS5jb206ZmFrZXBhc3N3b3Jk',
                  'Content-Type' => 'application/json',
                  'User-Agent' => 'The Best User Agent of All Time',
                  'X-App-Token' => 'totallyfakenotrealapptoken'
                })
          .to_return(resource('fakeuser.response'))

        response = @client.get('/api/users/current.json')
        assert_equal USER, response.email
      end
    end

    context 'RESTful updates' do
      should 'be able to POST a set of rows' do
        update = [{ :source => 'uw', :magnitude => 5, :earthquake_id => 42 }]
        stub_request(:post, 'https://fakehost.socrata.com/resource/earthquakes.json?')
          .with(
            :body => update.to_json,
            :headers => {
              'Accept' => '*/*',
              'Accept-Encoding' => 'gzip;q=1.0,deflate;q=0.6,identity;q=0.3',
              'Authorization' => 'Basic ZmFrZXVzZXJAc29jcmF0YS5jb206ZmFrZXBhc3N3b3Jk',
              'Content-Type' => 'application/json',
              'User-Agent' => 'The Best User Agent of All Time',
              'X-App-Token' => 'totallyfakenotrealapptoken'
            })
            .to_return(resource('earthquakes_create_one_row.response'))

        response = @client.post('earthquakes', update)
        assert_equal response['Rows Created'], 1
      end

      should 'be able to POST a form' do
        stub_request(:post, 'https://fakehost.socrata.com/resource/earthquakes.json?')
          .with(:body => { 'earthquake_id' => '42', 'magnitude' => '5', 'source' => 'uw' },
                :headers => { 'Accept' => '*/*', 
                              'Accept-Encoding' => 'gzip;q=1.0,deflate;q=0.6,identity;q=0.3',
                              'Authorization' => 'Basic ZmFrZXVzZXJAc29jcmF0YS5jb206ZmFrZXBhc3N3b3Jk',
                              'Content-Type' => 'application/x-www-form-urlencoded',
                              'User-Agent' => 'The Best User Agent of All Time',
                              'X-App-Token' => 'totallyfakenotrealapptoken' })
          .to_return(resource('earthquakes_create_one_row.response'))

        response = @client.post_form('earthquakes', :source => 'uw', :magnitude => 5, :earthquake_id => 42)
        assert_equal response['Rows Created'], 1
      end
    end
  end

  context 'errors' do
    setup do
      @client = SODA::Client.new(:domain => DOMAIN, :app_token => APP_TOKEN)
    end

    should 'get an error accessing a nonexistent dataset' do
      stub_request(:get, 'https://fakehost.socrata.com/resource/idontexist.json?')
        .to_return(resource('404.response'))

      assert_raise SODA::NotFound do
        @client.get('idontexist')
      end
    end
  end

  context 'raw' do
    setup do
      @client = SODA::Client.new(:domain => DOMAIN, :app_token => APP_TOKEN)
    end

    should 'be able to retrieve CSV if I so choose' do
      stub_request(:get, 'https://fakehost.socrata.com/resource/earthquakes.csv?')
        .to_return(resource('earthquakes.csv.response'))

      response = @client.get('earthquakes.csv')
      assert response.is_a? String
    end

    should 'be able to retrieve CSV with a full path' do
      stub_request(:get, 'https://fakehost.socrata.com/resource/earthquakes.csv?')
        .to_return(resource('earthquakes.csv.response'))

      response = @client.get('/resource/earthquakes.csv')
      assert response.is_a? String
    end
  end
end


lib/soda/exceptions.rb
# Adapted (lifted) from RestClient: https://github.com/rest-client/rest-client/blob/master/lib/restclient/exceptions.rb
module SODA
  STATUSES = {
    100 => 'Continue',
    101 => 'Switching Protocols',
    102 => 'Processing', # WebDAV

    200 => 'OK',
    201 => 'Created',
    202 => 'Accepted',
    203 => 'Non-Authoritative Information', # http/1.1
    204 => 'No Content',
    205 => 'Reset Content',
    206 => 'Partial Content',
    207 => 'Multi-Status', # WebDAV

    300 => 'Multiple Choices',
    301 => 'Moved Permanently',
    302 => 'Found',
    303 => 'See Other', # http/1.1
    304 => 'Not Modified',
    305 => 'Use Proxy', # http/1.1
    306 => 'Switch Proxy', # no longer used
    307 => 'Temporary Redirect', # http/1.1

    400 => 'Bad Request',
    401 => 'Unauthorized',
    402 => 'Payment Required',
    403 => 'Forbidden',
    404 => 'Not Found',
    405 => 'Method Not Allowed',
    406 => 'Not Acceptable',
    407 => 'Proxy Authentication Required',
    408 => 'Request Timeout',
    409 => 'Conflict',
    410 => 'Gone',
    411 => 'Length Required',
    412 => 'Precondition Failed',
    413 => 'Request Entity Too Large',
    414 => 'Request-URI Too Long',
    415 => 'Unsupported Media Type',
    416 => 'Requested Range Not Satisfiable',
    417 => 'Expectation Failed',
    418 => 'I\'m A Teapot', # RFC2324
    421 => 'Too Many Connections From This IP',
    422 => 'Unprocessable Entity', # WebDAV
    423 => 'Locked', # WebDAV
    424 => 'Failed Dependency', # WebDAV
    425 => 'Unordered Collection', # WebDAV
    426 => 'Upgrade Required',
    428 => 'Precondition Required', # RFC6585
    429 => 'Too Many Requests', # RFC6585
    431 => 'Request Header Fields Too Large', # RFC6585
    449 => 'Retry With', # Microsoft
    450 => 'Blocked By Windows Parental Controls', # Microsoft

    500 => 'Internal Server Error',
    501 => 'Not Implemented',
    502 => 'Bad Gateway',
    503 => 'Service Unavailable',
    504 => 'Gateway Timeout',
    505 => 'HTTP Version Not Supported',
    506 => 'Variant Also Negotiates',
    507 => 'Insufficient Storage', # WebDAV
    509 => 'Bandwidth Limit Exceeded', # Apache
    510 => 'Not Extended',
    511 => 'Network Authentication Required', # RFC6585
  }

  # This is the base exception class. Rescue it if you want to
  # catch any exception that your request might raise
  # You can get the status code by e.http_code, or see anything about the
  # response via e.response.
  # For example, the entire result body (which is
  # probably an HTML error page) is e.response.
  class Exception < RuntimeError
    attr_accessor :response
    attr_accessor :original_exception
    attr_accessor :data
    attr_writer :message

    def initialize(response = nil, initial_response_code = nil)
      @response = response
      @message = nil
      @initial_response_code = initial_response_code
    end

    def http_code
      # return integer for compatibility
      return @response.code.to_i if @response
      @initial_response_code
    end

    def http_headers
      @response.headers if @response
    end

    def http_body
      @response.body if @response
    end

    def to_s
      message
    end

    def message
      @message || default_message
    end

    def default_message
      self.class.name
    end
  end

  # Compatibility
  class ExceptionWithResponse < Exception
  end

  # The request failed with an error code not managed by the code
  class RequestFailed < ExceptionWithResponse
    def default_message
      "HTTP status code #{http_code}"
    end

    def to_s
      message
    end
  end

  # Exception classes. TODO: move all exceptions into this module.
  #
  # We will a create an exception for each status code, see
  # http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html
  #
  module Exceptions
    # Map http status codes to the corresponding exception class
    EXCEPTIONS_MAP = {}
  end

  # Create HTTP status exception classes
  STATUSES.each_pair do |code, message|
    klass = Class.new(RequestFailed) do
      send(:define_method, :default_message) { "#{http_code ? "#{http_code} " : ''}#{message}" }
    end
    klass_constant = const_set(message.delete(' \-\''), klass)
    Exceptions::EXCEPTIONS_MAP[code] = klass_constant
  end

  # Backwards compatibility. "Not Found" is the actual text in the RFCs.
  ResourceNotFound = NotFound

  module Exceptions
    # We have to split the Exceptions module like we do here because the
    # EXCEPTIONS_MAP is under Exceptions, but we depend on
    # SODA::RequestTimeout below.

    # Base class for request timeouts.
    #
    # NB: Previous releases of rest-client would raise RequestTimeout both for
    # HTTP 408 responses and for actual connection timeouts.
    class Timeout < SODA::RequestTimeout
      def initialize(message = nil, original_exception = nil)
        super(nil, nil)
        self.message = message if message
        self.original_exception = original_exception if original_exception
      end
    end

    # Timeout when connecting to a server. Typically wraps Net::OpenTimeout (in
    # ruby 2.0 or greater).
    class OpenTimeout < Timeout
      def default_message
        'Timed out connecting to server'
      end
    end

    # Timeout when reading from a server. Typically wraps Net::ReadTimeout (in
    # ruby 2.0 or greater).
    class ReadTimeout < Timeout
      def default_message
        'Timed out reading data from server'
      end
    end
  end

  # The server broke the connection prior to the request completing.  Usually
  # this means it crashed, or sometimes that your network connection was
  # severed before it could complete.
  class ServerBrokeConnection < Exception
    def initialize(message = 'Server broke connection')
      super nil, nil
      self.message = message
    end
  end

  class SSLCertificateNotVerified < Exception
    def initialize(message)
      super nil, nil
      self.message = message
    end
  end
end


lib/soda/client.rb
#!/usr/bin/env ruby
#
# A Ruby client to the RESTful Socrata Open Data API
#
# For more details, check out http://socrata.github.io/soda-ruby
#

require 'net/https'
require 'uri'
require 'json'
require 'cgi'
require 'hashie'
require 'sys/uname'
require 'soda/version'
require 'soda/exceptions'
include Sys

module SODA
  class Client
    class << self
      def generate_user_agent
        if Gem.win_platform?
          return "soda-ruby/#{SODA::VERSION} (Windows; Ruby/#{RUBY_VERSION}-p#{RUBY_PATCHLEVEL})"
        end
        "soda-ruby/#{SODA::VERSION} (#{Uname.uname.sysname}/#{Uname.uname.release}; Ruby/#{RUBY_VERSION}-p#{RUBY_PATCHLEVEL})"
      end
    end

    def blank?(object)
      object.nil? || object.empty?
    end

    ##
    #
    # Creates a new SODA client.
    #
    # * +config+ - A hash of the options to initialize the client with
    #
    # == Config Options
    #
    # * +:domain+ - The domain you want to access
    # * +:username+ - Your Socrata username (optional, only necessary for modifying data)
    # * +:password+ - Your Socrata password (optional, only necessary for modifying data)
    # * +:app_token+ - Your Socrata application token (register at http://dev.socrata.com/register)
    # * +:access_token+ - Your Socrata OAuth token (optional, https://dev.socrata.com/docs/authentication.html)
    # * +:ignore_ssl+ - Ignore SSL errors, which is very unsafe and only should be done in desperate circumstances (defaults to false)
    # * +:debug_stream+ - Set an output stream for debugging
    #
    # Returns a SODA::Client instance.
    #
    # == Example
    #
    #   client = SODA::Client.new({ :domain => "data.agency.gov", :app_token => "CGxarwoQlgQSev4zyUh5aR5J3" })
    #
    def initialize(config = {})
      @config = Hashie.symbolize_keys! config
      @user_agent = SODA::Client.generate_user_agent
    end

    ##
    #
    # Retrieve a resource via an HTTP GET request.
    #
    # * +resource+ - The resource identifier or path.
    # * +params+ - A hash of the URL parameters you want to pass
    #
    # Returns a Hashie::Mash that you can interact with as a Hash if you want.
    #
    # == Example
    #
    #   response = client.get("644b-gaut", { :firstname => "OPRAH", :lastname => "WINFREY", "$limit" => 5 })
    #
    def get(resource, params = {})
      connection(:get, resource, nil, params)
    end

    ##
    #
    # Update a resource via an HTTP POST request.
    #
    # Requires an authenticated client (both +:username+ and +:password+ passed into the +options+ hash)
    #
    # * +resource+ - The resource identifier or path.
    # * +body+ - The payload to POST. Will be converted to JSON (optional).
    # * +params+ - A hash of the URL parameters you want to pass (optional).
    #
    # Returns a Hashie::Mash that you can interact with as a Hash if you want.
    #
    # == Example
    #
    #   response = client.post("644b-gaut", [{ :firstname => "OPRAH", :lastname => "WINFREY" }])
    #
    def post(resource, body = nil, params = {})
      connection(:post, resource, body, params)
    end

    ##
    #
    # Replaces a resource via an HTTP PUT request.
    #
    # Requires an authenticated client (both +:username+ and +:password+ passed into the +options+ hash)
    #
    # * +resource+ - The resource identifier or path.
    # * +body+ - The payload to POST. Will be converted to JSON (optional).
    # * +params+ - A hash of the URL parameters you want to pass (optional).
    #
    # Returns a Hashie::Mash that you can interact with as a Hash if you want.
    #
    # == Example
    #
    #   response = client.put("644b-gaut", [{ :firstname => "OPRAH", :lastname => "WINFREY" }])
    #
    def put(resource, body = nil, params = {})
      connection(:put, resource, body, params)
    end

    ##
    #
    # Deletes a resource via an HTTP DELETE request.
    #
    # Requires an authenticated client (both +:username+ and +:password+ passed into the +options+ hash)
    #
    # * +resource+ - The resource identifier or path.
    # * +body+ - The payload to send with the DELETE. Will be converted to JSON (optional).
    # * +params+ - A hash of the URL parameters you want to pass (optional).
    #
    # Returns a Hashie::Mash that you can interact with as a Hash if you want.
    #
    # == Example
    #
    #   response = client.delete("644b-gaut")
    #
    def delete(resource, body = nil, params = {})
      connection(:delete, resource, body, params)
    end

    def post_form(resource, body = {}, params = {})
      # We'll combine any params we got from our base resource with
      # those passed in
      base = URI.parse(parse_resource(resource))
      query = [
        base.query,
        query_string(params)
      ].reject { |s| blank?(s) }.join '&'

      uri = URI.parse("https://#{base.host}#{base.path}?#{query}")

      request = Net::HTTP::Post.new(uri.request_uri)
      add_default_headers_to_request(request)
      request.set_form_data(body)

      # Authenticate if we're supposed to
      authenticate(request)

      # BAM!
      http = build_http_client(uri.host, uri.port)
      handle_response(http.request(request))
    end

    private

    def query_string(params)
      # Create query string of escaped key, value pairs
      params.map { |key, val| "#{key}=#{CGI.escape(val.to_s)}" }.join('&')
    end

    def parse_resource(resource)
      # If our resource starts with HTTPS, assume they've passed in a full URI
      return resource if resource.start_with?('https://')

      # If we didn't get a full path, assume "/resource/"
      resource = '/resource/' + resource unless resource.start_with?('/')

      # Check to see if we were given an output type
      extension = '.json'
      if matches = resource.match(/^(.+)(\.\w+)$/)
        resource = matches.captures[0]
        extension = matches.captures[1]
      end

      raise 'No base domain specified!' unless @config[:domain]
      "https://#{@config[:domain]}#{resource}#{extension}"
    end

    def handle_response(response)
      # Check our response code
      check_response_fail(response)
      if blank?(response.body)
        return nil
      elsif response['Content-Type'].include?('application/json')
        # Return a bunch of mashes if we're JSON
        response = JSON.parse(response.body, :max_nesting => false)
        if response.is_a? Array
          return response.map { |r| Hashie::Mash.new(r) }
        else
          return Hashie::Mash.new(response)
        end
      else
        # We don't partically care, just return the raw body
        return response.body
      end
    end

    def check_response_fail(response)
      return if %w(200 202).include? response.code

      # Adapted from RestClient's exception handling
      begin
        klass = SODA::Exceptions::EXCEPTIONS_MAP.fetch(response.code.to_i)

        raise klass.new(response, response.code)
      rescue KeyError
        raise RequestFailed.new(response, response.code)
      end
    end

    def connection(method = 'Get', resource = nil, body = nil, params = {})
      method = method.to_sym.capitalize

      # We'll combine any params we got from our base resource with
      # those passed in
      base = URI.parse(parse_resource(resource))
      query = [
        base.query,
        query_string(params)
      ].reject { |s| blank?(s) }.join '&'

      uri = URI.parse("https://#{base.host}#{base.path}?#{query}")

      request = eval("Net::HTTP::#{method}").new(uri.request_uri)
      add_default_headers_to_request(request)

      # Authenticate if we're supposed to
      authenticate(request)

      http = request_by_method(method, body, request, uri)
      send_request(method, http, request)
    end

    def send_request(method, http, request)
      return delete_method_response(http, request) if method === :Delete
      handle_response(http.request(request))
    end

    def delete_method_response(http, request)
      response = http.request(request)
      return response if response.code == '200'
      fail "Error querying \"#{uri}\": #{response.body}"
    end

    def net_http_class(method)
      Object.const_get("Net::HTTP::#{method}")
    end

    def request_by_method(method, body, request, uri)
      if [:Post, :Put, :Get].include?(method)
        request.content_type = 'application/json'
        request.body = body.to_json(:max_nesting => false)
      end

      build_http_client(uri.host, uri.port)
    end

    def build_http_client(host, port)
      http = Net::HTTP.new(host, port)
      http.use_ssl = true
      http.verify_mode = OpenSSL::SSL::VERIFY_NONE if @config[:ignore_ssl]
      http.read_timeout = @config[:timeout] if @config[:timeout]
      http.set_debug_output(@config[:debug_stream]) if @config[:debug_stream]
      http
    end

    def authenticate(request)
      # Authenticate if we're supposed to
      if @config[:username]
        request.basic_auth @config[:username], @config[:password]
      elsif @config[:access_token]
        request.add_field('Authorization', "OAuth #{@config[:access_token]}")
      end
    end

    def add_default_headers_to_request(request)
      request.delete('User-Agent')
      request.add_field('X-App-Token', @config[:app_token])
      request.add_field('User-Agent', @user_agent)
    end
  end
end
